File: matlab/src/api/pfqn/pfqn_mcub.m
%{
%{
 % @file pfqn_mcub.m
 % @brief Multiclass Composite Upper Bound (Kerola 1986) on per-class
 %        throughput for closed product-form networks.
%}
%}

function [Xub,Xlb] = pfqn_mcub(L,N,Z)
%{
%{
 % @brief Kerola's composite bound method (Perf. Eval. 6:1-9, eqs. 10-16).
 %        Given per-class multiclass BJB lower bounds X_s^-, the residual-
 %        utilization composite upper bound is
 %          X_r <= min_k [1 - sum_{s!=r} X_s^- L_ks] / L_kr,
 %        computed at O(KR). Much tighter than per-class ABA at moderate load.
 %        (Named pfqn_mcub because pfqn_cub is the unrelated cubature NC method.)
 % @fn pfqn_mcub(L, N, Z)
 % @param L Service demand matrix, station x class (M x R).
 % @param N Population vector (1 x R).
 % @param Z Think time vector (1 x R, default zeros).
 % @return Xub Composite upper throughput bound per class (1 x R).
 % @return Xlb Multiclass BJB lower throughput bound per class (1 x R, eq. 10).
%}
%}
[M,R] = size(L);
N = N(:)';
if nargin < 3 || isempty(Z), Z = zeros(1,R); end
Z = Z(:)';
Ntot = sum(N);
R0 = sum(L,1);                       % per-class total demand (1 x R)
Lb = max(L,[],1);                    % per-class bottleneck demand (1 x R)

% eq (10): multiclass Balanced Job Bounds lower throughput bound.
Xlb = N ./ (R0 + Z + (Ntot - 1).*Lb);

% eqs (13)-(16): composite upper bound per class.
Xub = zeros(1,R);
for r = 1:R
    Uoth = zeros(M,1);              % utilization at each device by other classes
    for s = 1:R
        if s ~= r
            Uoth = Uoth + Xlb(s) * L(:,s);
        end
    end
    Ucub = 1 - Uoth;               % residual utilization available to class r
    dev = inf(M,1);
    for k = 1:M
        if L(k,r) > 0
            dev(k) = Ucub(k) / L(k,r);
        end
    end
    Xub(r) = min(dev);
end
end

File: jar/src/main/java/jline/api/pfqn/mva/Pfqn_mcub.java
/**
 * @file Multiclass Composite Upper Bound (Kerola 1986)
 *
 * Kerola's composite bound method (Perf. Eval. 6:1-9, eqs. 10-16): a multiclass BJB lower
 * bound (eq. 10) seeds a residual-utilization composite upper bound (eqs. 13-16). Named
 * mcub because pfqn_cub is the unrelated cubature NC method. Ported at parity from MATLAB
 * pfqn_mcub.m.
 *
 * @since LINE 3.0
 */
package jline.api.pfqn.mva;

import jline.util.matrix.Matrix;

public final class Pfqn_mcub {
    private Pfqn_mcub() {}

    /**
     * Multiclass composite upper bound and its BJB lower seed.
     *
     * @param L service demand matrix, station x class (M x R)
     * @param N population vector (1 x R)
     * @param Z think time vector (1 x R)
     * @return {Xub, Xlb} 1xR upper (composite) and lower (multiclass BJB) throughput bounds
     */
    public static Matrix[] pfqn_mcub(Matrix L, Matrix N, Matrix Z) {
        int M = L.getNumRows();
        int R = L.getNumCols();
        double Ntot = N.elementSum();
        Matrix Xlb = new Matrix(1, R);
        Matrix Xub = new Matrix(1, R);
        // eq (10): multiclass Balanced Job Bounds lower throughput bound.
        for (int r = 0; r < R; r++) {
            double R0 = 0.0, Lb = 0.0;
            for (int k = 0; k < M; k++) {
                R0 += L.get(k, r);
                if (L.get(k, r) > Lb) Lb = L.get(k, r);
            }
            double zr = (Z != null && Z.getNumCols() > r) ? Z.get(0, r) : 0.0;
            Xlb.set(0, r, N.get(0, r) / (R0 + zr + (Ntot - 1) * Lb));
        }
        // eqs (13)-(16): composite upper bound per class.
        for (int r = 0; r < R; r++) {
            double best = Double.POSITIVE_INFINITY;
            for (int k = 0; k < M; k++) {
                if (L.get(k, r) > 0) {
                    double uoth = 0.0;
                    for (int s = 0; s < R; s++) {
                        if (s != r) uoth += Xlb.get(0, s) * L.get(k, s);
                    }
                    double dev = (1.0 - uoth) / L.get(k, r);
                    if (dev < best) best = dev;
                }
            }
            Xub.set(0, r, best);
        }
        return new Matrix[]{Xub, Xlb};
    }
}
